### PR TITLE
Issue 7281 on CodePlex: Unable to receive SNMPv3 traps

### PR DESCRIPTION
**Pull request by KristianVirkus**
*Makes authenticated SNMPv3 TrapV2Pdus pass instead of insisting on Reportable message header flag.
*Modified IsInTime(...) method for timeliness check of received SNMPv3 messages.

### CODE REVIEW

File: SharpSnmpLib/Pipeline/EngineGroup.cs [hunk 1]
```diff
@@ -111,8 +111,9 @@ public static bool IsInTime(int[] currentTimeData, int pastReboots, int pastTime
                 return false;
             }
 
-            var diff = currentTimeData[1] > pastTime ? currentTimeData[1] - pastTime : currentTimeData[1] - pastTime - int.MinValue + int.MaxValue;
```
Comment (KristianVirkus) [inline SharpSnmpLib/Pipeline/EngineGroup.cs]: What's the idea behind this expression? At least in the related unit test it leads to notInTimeWindow responses no matter which positive engine times I create the TrapV2Message object with.

File: SharpSnmpLib/Pipeline/SecureSnmpContext.cs [hunk 2]
```diff
@@ -189,7 +189,8 @@ public override bool HandleMembership()
                 return false;
             }
 
-            if ((user.ToSecurityLevel() | Levels.Reportable) != request.Header.SecurityLevel)
+			// Require report flag to be set for all PDUs but TrapV2Pdu (see RFC 3412, section 6.4.)
+			if ((request.Pdu().GetType() != typeof(TrapV2Pdu)) && ((user.ToSecurityLevel() | Levels.Reportable) != request.Header.SecurityLevel))
```
Comment (KristianVirkus) [inline SharpSnmpLib/Pipeline/SecureSnmpContext.cs]: According to RFC 3412 there are some other cases which don't require a response. But at least in section 6.4 only some and "Unconfirmed Class" PDUs were mentioned. Apart from that I'm not sure which parts of the #SNMP library use the SecureSnmpContext and which effects changes could have. Maybe even this change cannot be used.

File: Tests/Pipeline/Tests/SecureSnmpContextTestFixture.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			SecureSnmpContext context;
+			this.setup1(out trapMessage, out context);
+
+			// Enforce Reportable flag.
+			// TODO Replace string literal by nameof(...) as soon as switched to a more recent version of C#.
+			typeof(Header).GetProperty("SecurityLevel", BindingFlags.Public | BindingFlags.Instance).SetValue(trapMessage.Header, trapMessage.Header.SecurityLevel | Levels.Reportable);
+
+			// Accept either true as return value or a "not in time window" reponse.
+			var result = context.HandleMembership();
+			Assert.IsTrue(result);
```
Comment (KristianVirkus) [inline Tests/Pipeline/Tests/SecureSnmpContextTestFixture.cs]: This is somehow superfluous. It is handled by the next line. This was for debugging.......

**Comment by lextm** [comment]
Should simply let TRAP v2 messages pass authentication. Closed this PR and use c55d7a46eec74bf78182b8085af7f6e6317374e0 instead.

**Pull request closed without merging**